File: libs/python/alp_flags/src/alp_flags/decision_log.py
"""Decision-log helper — turns the alpflags `on_decision` hook into a
`flag.decision` structlog event with consistent fields across all services.

Usage in a service's connect_flags():

    from alp_flags import FlagClient
    from alp_flags.decision_log import structlog_decision_hook

    _client = FlagClient(
        institution_url=...,
        on_decision=structlog_decision_hook("auth"),
        ...
    )

The emitted event:

    {
      "event": "flag.decision",
      "service": "auth",
      "flag_name": "email_channel_enabled",
      "tenant_id": null,
      "value": true,
      "source": "cache",
      "fallback_reason": null
    }

Source labels: "cache" (hot path), "institution" (HTTP fetch), "fallback"
(institution unhealthy or unknown flag — `fallback_reason` carries detail).
A spike in source=fallback is the standard signal that Institution is sick.
"""

from __future__ import annotations

from collections.abc import Awaitable, Callable
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from alp_flags.client import Decision


def structlog_decision_hook(service_name: str) -> Callable[["Decision"], Awaitable[None]]:
    """Return an OnDecision callback that emits structlog `flag.decision` events.

    Imports structlog lazily so the alp_flags lib stays usable in services
    that haven't installed structlog yet (none today, but defensive).
    """
    import structlog

    logger = structlog.get_logger("alp_flags.decision")

    async def _hook(d: "Decision") -> None:
        logger.info(
            "flag.decision",
            service=service_name,
            flag_name=d.flag_name,
            tenant_id=d.tenant_id,
            value=d.value,
            source=d.source,
            fallback_reason=d.fallback_reason,
        )

    return _hook

File: libs/go/alpflags/decision_log.go
package alpflags

import "log/slog"

// SlogDecisionHook returns an OnDecision callback that emits one slog INFO
// event per evaluation, with the same field names used by the Python sibling
// (libs/python/alp_flags/decision_log.py) so logs across the platform have
// a single shape:
//
//	{
//	  "level": "INFO",
//	  "msg":   "flag.decision",
//	  "service":         "<service-name>",
//	  "flag_name":       "<flag>",
//	  "tenant_id":       "<tenant or empty>",
//	  "value":           true,
//	  "source":          "cache" | "institution" | "fallback",
//	  "fallback_reason": "" | "institution_error:..."
//	}
//
// Pass nil logger to use slog.Default(). Service name is required so cross-
// service queries can filter by it.
func SlogDecisionHook(serviceName string, logger *slog.Logger) func(Decision) {
	if logger == nil {
		logger = slog.Default()
	}
	return func(d Decision) {
		logger.Info("flag.decision",
			"service", serviceName,
			"flag_name", d.FlagName,
			"tenant_id", d.TenantID,
			"value", d.Value,
			"source", d.Source,
			"fallback_reason", d.FallbackReason,
		)
	}
}
